File: calendar/calendar.m
#import <Foundation/Foundation.h>
#import <stdio.h>

@interface Calendar: NSObject

- (NSArray *) makeMonth: (int) month inYear: (int) year;

@end

@implementation Calendar

- (NSArray *) makeMonth: (int) month inYear: (int) year {
  NSDateComponents *monthComps = [[NSDateComponents alloc] init];
  NSDateComponents *lastComps = [[NSDateComponents alloc] init];
  NSDateComponents *yesterdayComps = [[NSDateComponents alloc] init];
  NSDateComponents *tomorrowComps = [[NSDateComponents alloc] init];

  monthComps.year = year;
  monthComps.month = month;
  lastComps.month = 1;
  lastComps.day = -1;
  yesterdayComps.day = -1;
  tomorrowComps.day = 1;
  
  NSCalendar *calendar = [NSCalendar currentCalendar];
  NSMutableArray *weeks = [NSMutableArray array];
  NSDate *date = [calendar dateFromComponents:monthComps];
  NSDate *last = [calendar dateByAddingComponents:lastComps toDate:date options:0];
  
  while ([calendar components:NSCalendarUnitWeekday fromDate: date].weekday > 1) {
    date = [calendar dateByAddingComponents:yesterdayComps toDate:date options:0];
  }
  
  while ([date compare: last] == NSOrderedAscending) {
    NSMutableArray *week = [NSMutableArray array];
    for (int i = 0; i < 7; i++) {
      NSInteger day = [calendar components:NSCalendarUnitDay fromDate:date].day;
      [week addObject: [NSNumber numberWithInt: day]];
      date = [calendar dateByAddingComponents:tomorrowComps toDate:date options:0];
    }
    [weeks addObject: week];
  }
  
  return weeks;
}

@end

int main (int argc, const char * argv[])
{
  @autoreleasepool {
    Calendar *calendar = [[Calendar alloc] init];
    NSArray *weeks = [calendar makeMonth:12 inYear:2021];
    
    for (NSArray *week in weeks) {
      NSLog(@"%@", week);
    }

    if ([[calendar makeMonth:2 inYear:2015] count] == 4) {
      NSLog(@"Feb 2015 has 4 weeks");
    }
    if ([[calendar makeMonth:4 inYear:2021] count] == 5) {
      NSLog(@"Apr 2021 has 5 weeks");
    }
    if ([[calendar makeMonth:5 inYear:2021] count] == 6) {
      NSLog(@"May 2015 has 6 weeks");
    }
  }
}

File: calendar/calendar.swift
import Foundation

let calendar = Calendar(identifier: .gregorian)

func makeMonth(month: Int, year: Int) -> [[Int]] {
  var weeks = [[Int]]()
  var date = calendar.date(from: DateComponents(year: year, month: month))!
  let last = calendar.date(byAdding: DateComponents(month: 1, day: -1), to: date)!
  
  while calendar.component(.weekday, from: date) > 1 {
    date = calendar.date(byAdding: .day, value: -1, to: date, wrappingComponents: false)!
  }
  while date <= last {
    var week = [Int]()
    for _ in 1...7 {
      week.append(calendar.component(.day, from: date))
      date = calendar.date(byAdding: .day, value: 1, to: date, wrappingComponents: false)!
    }
    weeks.append(week)
  }
  return weeks
}

let weeks = makeMonth(month: 12, year: 2021)
for week in weeks {
  print("\(week)")
}

if makeMonth(month: 2, year: 2015).count == 4 {
  print("Feb 2015 has 4 weeks")
}
if makeMonth(month: 4, year: 2021).count == 5 {
  print("Apr 2021 has 5 weeks")
}
if makeMonth(month: 5, year: 2021).count == 6 {
  print("May 2021 has 6 weeks")
}
